<a href="https://colab.research.google.com/github/hjhan1201/AIFFEL_Online_Quest/blob/main/DeepLearning_1S1K_Tensor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [56]:
import tensorflow as tf

In [57]:
t0 = tf.constant(1)
print(t0)
print(tf.rank(t0)) # 축의 개수를 반환합니다.

tf.Tensor(1, shape=(), dtype=int32)
tf.Tensor(0, shape=(), dtype=int32)


In [58]:
t1 = tf.constant([1, 2, 3])
print(t1)
print(tf.rank(t1))

tf.Tensor([1 2 3], shape=(3,), dtype=int32)
tf.Tensor(1, shape=(), dtype=int32)


In [59]:
# Q. 위 이미지와 같은 크기 (3, 3)의 2차원 텐서를 만들고, 차원을 반환해 주세요.
t2 = tf.constant([[1, 2, 3],[1, 2, 3],[1, 2, 3]])
print(t2)
print(tf.rank(t2))

tf.Tensor(
[[1 2 3]
 [1 2 3]
 [1 2 3]], shape=(3, 3), dtype=int32)
tf.Tensor(2, shape=(), dtype=int32)


In [60]:
# Q. 위 이미지와 같은 크기(3, 3, 3)의 3차원 텐서를 만들고, 차원을 반환해 주세요.
t3 = tf.constant([[[1, 2, 3],[1, 2, 3],[1, 2, 3]],[[1, 2, 3],[1, 2, 3],[1, 2, 3]],[[1, 2, 3],[1, 2, 3],[1, 2, 3]]])
print(t3)
print(tf.rank(t3))

tf.Tensor(
[[[1 2 3]
  [1 2 3]
  [1 2 3]]

 [[1 2 3]
  [1 2 3]
  [1 2 3]]

 [[1 2 3]
  [1 2 3]
  [1 2 3]]], shape=(3, 3, 3), dtype=int32)
tf.Tensor(3, shape=(), dtype=int32)


In [61]:
i = tf.constant(2)
print(i)

tf.Tensor(2, shape=(), dtype=int32)


In [62]:
f = tf.constant(2.)
print(f)

tf.Tensor(2.0, shape=(), dtype=float32)


In [63]:
s = tf.constant('Suan')
print(s)

tf.Tensor(b'Suan', shape=(), dtype=string)


In [64]:
f16 = tf.constant(2., dtype=tf.float16)
print(f16)

tf.Tensor(2.0, shape=(), dtype=float16)


In [65]:
# Q. 8비트의 정수형을 지정하여 텐서를 생성해 보세요.
f8 = tf.constant(2, dtype=tf.int8)
print(f8)

tf.Tensor(2, shape=(), dtype=int8)


In [66]:
f32 = tf.cast(f16, tf.float32)
print(f32)

tf.Tensor(2.0, shape=(), dtype=float32)


In [67]:
#  Q. 8비트 정수형을 32비트 정수형으로 변환해 보세요.
f322 = tf.cast(f, tf.int32)

In [68]:
x = tf.constant([[1], [2], [3]])
print(x)
print(x.shape)

y = tf.reshape(x, [1, 3])
print(y)
print(y.shape)

tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
(3, 1)
tf.Tensor([[1 2 3]], shape=(1, 3), dtype=int32)
(1, 3)


In [69]:
print(y)
print(tf.transpose(y))
print(y.shape)

tf.Tensor([[1 2 3]], shape=(1, 3), dtype=int32)
tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
(1, 3)


In [70]:
print(x)
print(tf.squeeze(x))

tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
tf.Tensor([1 2 3], shape=(3,), dtype=int32)


In [71]:
print(y)
print(tf.expand_dims(y, axis=0))
print(tf.expand_dims(y, axis=1))
print(tf.expand_dims(y, axis=2))

tf.Tensor([[1 2 3]], shape=(1, 3), dtype=int32)
tf.Tensor([[[1 2 3]]], shape=(1, 1, 3), dtype=int32)
tf.Tensor([[[1 2 3]]], shape=(1, 1, 3), dtype=int32)
tf.Tensor(
[[[1]
  [2]
  [3]]], shape=(1, 3, 1), dtype=int32)


In [72]:
print(x)
print(tf.split(x, 3))

tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
[<tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[1]], dtype=int32)>, <tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[2]], dtype=int32)>, <tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[3]], dtype=int32)>]


In [73]:
print(x)
print(tf.concat([x, x], axis=0))
print(tf.concat([x, x], axis=1))

tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
tf.Tensor(
[[1]
 [2]
 [3]
 [1]
 [2]
 [3]], shape=(6, 1), dtype=int32)
tf.Tensor(
[[1 1]
 [2 2]
 [3 3]], shape=(3, 2), dtype=int32)


In [74]:
print(tf.constant(2) + tf.constant(2))
print(tf.constant(2) - tf.constant(2))
print(tf.add(tf.constant(2), tf.constant(2)))
print(tf.subtract(tf.constant(2), tf.constant(2)))

tf.Tensor(4, shape=(), dtype=int32)
tf.Tensor(0, shape=(), dtype=int32)
tf.Tensor(4, shape=(), dtype=int32)
tf.Tensor(0, shape=(), dtype=int32)


In [75]:
print(tf.constant(2) * tf.constant(2))
print(tf.constant(2) / tf.constant(2))
print(tf.multiply(tf.constant(2), tf.constant(2)))
print(tf.divide(tf.constant(2), tf.constant(2)))

tf.Tensor(4, shape=(), dtype=int32)
tf.Tensor(1.0, shape=(), dtype=float64)
tf.Tensor(4, shape=(), dtype=int32)
tf.Tensor(1.0, shape=(), dtype=float64)


In [76]:
#print(tf.constant(2) + tf.constant(2.2))

텐서 연산에서 주의할 점은 서로 다른 타입을 가지는 텐서는 연산이 되지 않고, 에러가 발생한다는 점입니다.

다른 타입을 가지는 텐서는 연산을 위해서 타입을 변환해 주어야합니다. 2-3 스텝에서 언급했던 tf.cast를 사용하여 텐서의 타입을 서로 맞춰준 후에 연산을 할 수 있습니다

In [77]:
print(tf.cast(tf.constant(2), tf.float32) + tf.constant(2.2))

tf.Tensor(4.2, shape=(), dtype=float32)


In [78]:
# Q. 1차원 텐서 2개를 생성해 보세요.
t1 = tf.constant([1, 2, 3])
t2 = tf.constant([4, 5, 6])

In [79]:
# Q. 사칙연산(더하기, 빼기, 곱하기, 나누기)을 해보세요.
print(t1+t2)
print(t1-t2)
print(t1*t2)
print(t1/t2)

tf.Tensor([5 7 9], shape=(3,), dtype=int32)
tf.Tensor([-3 -3 -3], shape=(3,), dtype=int32)
tf.Tensor([ 4 10 18], shape=(3,), dtype=int32)
tf.Tensor([0.25 0.4  0.5 ], shape=(3,), dtype=float64)


In [80]:
# Q. 크기 (2,2)인 2차원 텐서 2개를 생성해 보세요.
a = tf.constant([[1, 2],
                [3,4]])
b = tf.constant([[5, 6],
                [7,8]])

In [81]:
print(a + b) # element-wise addition
print(a - b) # element-wise subtraction
print(a * b) # element-wise multiplication
print(a @ b) # matrix multiplication
print(a / b) # element-wise division

tf.Tensor(
[[ 6  8]
 [10 12]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[-4 -4]
 [-4 -4]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[ 5 12]
 [21 32]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[19 22]
 [43 50]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[0.2        0.33333333]
 [0.42857143 0.5       ]], shape=(2, 2), dtype=float64)


In [82]:
print(tf.add(a, b))
print(tf.subtract(a, b))
print(tf.multiply(a, b))
print(tf.matmul(a, b))
print(tf.divide(a, b))

tf.Tensor(
[[ 6  8]
 [10 12]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[-4 -4]
 [-4 -4]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[ 5 12]
 [21 32]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[19 22]
 [43 50]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[0.2        0.33333333]
 [0.42857143 0.5       ]], shape=(2, 2), dtype=float64)


In [83]:
c = tf.constant([[4.0, 5.0, 6.0],
                 [10.0, 9.0, 8.0]])

print(tf.reduce_max(c))
print(tf.argmax(c))
print(tf.nn.softmax(c))

tf.Tensor(10.0, shape=(), dtype=float32)
tf.Tensor([1 1 1], shape=(3,), dtype=int64)
tf.Tensor(
[[0.09003057 0.24472848 0.66524094]
 [0.66524094 0.24472848 0.09003057]], shape=(2, 3), dtype=float32)


reduce_max(): 텐서 값 중에서 최대값을 계산합니다.

argmax(): 최대값의 위치를 반환합니다. 예제에서는 [4.0, 5.0, 6.0] 보다 [10.0, 9.0, 8.0]이 다 크기 때문에 [0 0 0]이 아니라 [1 1 1]의 결과를 보여주고 있습니다.

nn.softmax(): 텐서의 값을 0과 1 사이의 값으로 보여줍니다. 예제에서는 [4.0, 5.0, 6.0] 값이 [0.09003057 0.24472848 0.66524094]으로 변환되었고, [10.0, 9.0, 8.0]이 [0.66524094 0.24472848 0.09003057]로 변환되었습니다.

---



In [84]:
# [Playground] 다양한 텐서의 연산을 연습해 보세요!
t2 = tf.constant([[1, 2, 3],[4, 5, 6],[7, 8, 9]])
t4 = tf.constant([[1, 2, 3],[1, 2, 3],[1, 2, 3]])

print(t2 +t4) # element-wise addition
print(t2 - t4) # element-wise subtraction
print(t2 * t4) # element-wise multiplication
print(t2 @ t4) # matrix multiplication
print(t2 / t4) # element-wise division

tf.Tensor(
[[ 2  4  6]
 [ 5  7  9]
 [ 8 10 12]], shape=(3, 3), dtype=int32)
tf.Tensor(
[[0 0 0]
 [3 3 3]
 [6 6 6]], shape=(3, 3), dtype=int32)
tf.Tensor(
[[ 1  4  9]
 [ 4 10 18]
 [ 7 16 27]], shape=(3, 3), dtype=int32)
tf.Tensor(
[[ 6 12 18]
 [15 30 45]
 [24 48 72]], shape=(3, 3), dtype=int32)
tf.Tensor(
[[1.  1.  1. ]
 [4.  2.5 2. ]
 [7.  4.  3. ]], shape=(3, 3), dtype=float64)


In [85]:
a = tf.constant([1, 2, 3, 4, 5, 6, 7, 8])
b = tf.constant([[1., 2., 3., 4., 5., 6., 7., 8.]])

# Q. 위에 있는 2개의 텐서를 행렬곱연산이 가능하도록 3차원 텐서로 변환해주세요.
a = tf.reshape(a,[2,2,2])
b = tf.transpose(tf.reshape(b, [2, 2, 2]))
b = tf.cast(b, tf.int32)
print(a)
print(b)

print(a@b)

tf.Tensor(
[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]], shape=(2, 2, 2), dtype=int32)
tf.Tensor(
[[[1 5]
  [3 7]]

 [[2 6]
  [4 8]]], shape=(2, 2, 2), dtype=int32)
tf.Tensor(
[[[  7  19]
  [ 15  43]]

 [[ 34  78]
  [ 46 106]]], shape=(2, 2, 2), dtype=int32)
